app/javascript/dashboard/helper/messageHelper.js

```diff
@@ -0,0 +1,63 @@
+const MESSAGE_VARIABLES_REGEX = /{{(.*?)}}/g;
+export const replaceVariablesInMessage = ({ message, variables }) => {
+  return message.replace(MESSAGE_VARIABLES_REGEX, (match, replace) => {
+    return variables[replace.trim()]
+      ? variables[replace.trim().toLowerCase()]
+      : '';
+  });
+};
+
+const skipCodeBlocks = str => str.replace(/```(?:.|\n)+?```/g, '');
+
+export const getFirstName = ({ name }) => {
+  return name.split(' ')[0];
+};
+
+export const getLastName = ({ name }) => {
+  return name.split(' ').length > 1
+    ? name.split(' ')[name.split(' ').length - 1]
+    : '';
+};
+
+export const getMessageVariables = ({ conversation }) => {
+  const {
+    meta: { assignee = {}, sender = {} },
+    id,
+  } = conversation;
+
+  return {
+    'contact.name': sender?.name,
+    'contact.first_name': getFirstName({ name: sender?.name }),
+    'contact.last_name': getLastName({ name: sender?.name }),
+    'contact.email': sender?.email,
+    'contact.phone': sender?.phone_number,
+    'conversation.id': id,
+    'agent.name': assignee?.name ? assignee?.name : '',
+    'agent.first_name': assignee?.name
+      ? getFirstName({ name: assignee?.name })
+      : '',
+    'agent.last_name': assignee?.name
+      ? getLastName({ name: assignee?.name })
+      : '',
+    'agent.email': assignee?.email ? assignee?.email : '',
+  };
+};
+
+export const getUndefinedVariablesInMessage = ({ message, variables }) => {
+  const messageWithOutCodeBlocks = skipCodeBlocks(message);
+  const matches = messageWithOutCodeBlocks.match(MESSAGE_VARIABLES_REGEX);
+  const undefinedVariables = [];
+  if (!matches) {
+    return [];
+  }
+  matches.forEach(match => {
+    const variable = match
+      .replace('{{', '')
+      .replace('}}', '')
+      .trim();
+    if (!variables[variable]) {
+      undefinedVariables.push(match);
+    }
+  });
+  return undefinedVariables;
+};
```

Comment:
We can make a couple of changes here

1. The declaration of `undefinedVariables` can be after the check for the `!matches` on line 50. 
2. The `forEach` can be easily replaced with a `reduce`
3. In case the forEach is not very readable we can use `map` and `filter`, while it does two passes at the array it does avoid declaring `undefinedVariables` and push to it in the first thing.
4. We can replace the `match.replace('{{', '')` with a regex match like `const [, variable] = match.match(EXTRACTION_REGEX) || [];` we can have the extraction regex as `/{{\s*(.+?)\s*}}/` This will do exactly what we need

Feel free to make changes (or not to make them) as you see fit, here's a sample of what I found to be the most readable

```suggestion
export const getUndefinedVariablesInMessage = ({ message, variables }) => {
  const messageWithOutCodeBlocks = skipCodeBlocks(message);
  const matches = messageWithOutCodeBlocks.match(MESSAGE_VARIABLES_REGEX);
  if (!matches) return [];

  return matches.map(match => {
    return match.replace('{{', '').replace('}}', '').trim()
  }).filter(variable => {
    return !variables[variable]
  })
};
```